docs/programmers_guide/witness_formal_spec.md

```diff
@@ -17,13 +17,14 @@ In the end, when there are no more instructions left, there MUST be only one ite
 
 ### Nodes
 
-There are multiple types of nodes that can be present on the stack.
-
-- `hashNode`
-- `branchNode`
-- `accountNode`
-- `valueNode`
-- `extensionNode`
+```
+type Node = nil
+          | HashNode (raw_hash:Byte[32])
+          | ValueNode (raw_value:Byte[])
+          | AccountNode (nonce:Uint balance:int storage:Node storage_hash:byte[32] code_hash:byte[32])
+          | ShortNode (key:Byte[] value:node)
```

Comment:
I would separate the cases of leaf nodes and extension nodes. For two reasons. First, "ShortNode" is not a very helpful name. Second, separating these two cases would eliminate the case where ShortNode's value is also a short node. Otherwise you would need to introduce some kind of type predicates, which is more complex than we want

Reply:
leaf node is already introduced (probably should rename it from `ValueNode`) so basically we need to change ShortNode to Extension or something like that.

Reply:
ah, no, you are right, it should be `Leaf{key, ValueNode}` or `Leaf{key, AccountNode}` and `Extension{key, Leaf{key2, ValueNode}}`